src/librustc_lint/builtin.rs

```diff
@@ -1876,25 +1876,103 @@ declare_lint_pass!(InvalidValue => [INVALID_VALUE]);
+            const ZEROED_PATH: &[Symbol] = &[sym::core, sym::mem, sym::zeroed];
+            const UININIT_PATH: &[Symbol] = &[sym::core, sym::mem, sym::uninitialized];
+            // `transmute` is inside an anonymous module (the `extern` block?);
+            // `Invalid` represents the empty string and matches that.
+            const TRANSMUTE_PATH: &[Symbol] =
+                &[sym::core, sym::intrinsics, kw::Invalid, sym::transmute];
+
+            if let hir::ExprKind::Call(ref path_expr, ref args) = expr.node {
+                if let hir::ExprKind::Path(ref qpath) = path_expr.node {
+                    if let Some(def_id) = cx.tables.qpath_res(qpath, path_expr.hir_id)
+                        .opt_def_id()
+                    {
+                        if cx.match_def_path(def_id, &ZEROED_PATH) {
+                            return Some(InitKind::Zeroed);
+                        }
+                        if cx.match_def_path(def_id, &UININIT_PATH) {
+                            return Some(InitKind::Uninit);
+                        }
+                        if cx.match_def_path(def_id, &TRANSMUTE_PATH) {
+                            if is_zero(&args[0]) {
+                                return Some(InitKind::Zeroed);
+                            }
+                        }
+                        // FIXME: Also detect `MaybeUninit::zeroed().assume_init()` and
+                        // `MaybeUninit::uninit().assume_init()`.
+                    }
+                }
+            }
+
+            None
+        }
+
         /// Return `Some` only if we are sure this type does *not*
         /// allow zero initialization.
-        fn ty_find_init_error<'tcx>(tcx: TyCtxt<'tcx>, ty: Ty<'tcx>) -> Option<InitError> {
+        fn ty_find_init_error<'tcx>(
+            tcx: TyCtxt<'tcx>,
+            ty: Ty<'tcx>,
+            init: InitKind,
+        ) -> Option<InitError> {
             use rustc::ty::TyKind::*;
             match ty.sty {
                 // Primitive types that don't like 0 as a value.
                 Ref(..) => Some((format!("References must be non-null"), None)),
                 Adt(..) if ty.is_box() => Some((format!("`Box` must be non-null"), None)),
                 FnPtr(..) => Some((format!("Function pointers must be non-null"), None)),
                 Never => Some((format!("The never type (`!`) has no valid value"), None)),
-                // Recurse for some compound types.
+                // Primitive types with other constraints
+                Bool if init == InitKind::Uninit =>
+                    Some((format!("Booleans must be `true` or `false`"), None)),
+                Char if init == InitKind::Uninit =>
+                    Some((format!("Characters must be a valid unicode codepoint"), None)),
+                // Recurse and checks for some compound types.
                 Adt(adt_def, substs) if !adt_def.is_union() => {
```

Comment:
Consider moving parts (the two large sections) or the whole arm into its own function; it's getting a bit long.

Reply:
TBH I feel this is still reasonable. The individual parts as their own function don't feel very self-contained.

Reply:
Alright :)